examples/vimco.py

```diff
@@ -0,0 +1,385 @@
+
+(LL_eval,
+ log_qz_given_x_eval,
+ log_pz_eval,
+ log_px_given_z_eval,
+ _) = lower_bound(
+    z_eval, z_mu_eval,
+    x_mu_eval, sym_x,
+    eq_samples=sym_eq_samples, iw_samples=sym_iw_samples)
+
+#some sanity checks that we can forward data through the model
+# dummy data for testing the implementation
+X = np.ones((batch_size, 784), dtype=theano.config.floatX)
+
+print "OUTPUT SIZE OF l_z1 using BS=%d, latent_size=%d, sym_iw_samples=%d, sym_eq_samples=%d --"\
+      %(batch_size, latent_size, iw_samples, eq_samples), \
+    lasagne.layers.get_output(l_z,sym_x).eval(
+    {sym_x: X, sym_iw_samples: np.int32(iw_samples),
+     sym_eq_samples: np.int32(eq_samples)}).shape
+
+params = p_params + q_params
+
+clip_grad = 1
+max_norm = 5
+mgrads = lasagne.updates.total_norm_constraint(grads,max_norm=max_norm)
+cgrads = [T.clip(g, -clip_grad, clip_grad) for g in mgrads]
+
+updates = lasagne.updates.adam(
+    cgrads, params, beta1=0.9, beta2=0.999, epsilon=1e-4, learning_rate=sym_lr)
+
+# Helper symbolic variables to index into the shared train and test data
+sym_index = T.iscalar('index')
+sym_batch_size = T.iscalar('batch_size')
+batch_slice = slice(sym_index * sym_batch_size, (sym_index + 1) * sym_batch_size)
+
+train_model = theano.function(
+    [sym_index, sym_batch_size, sym_lr, sym_eq_samples, sym_iw_samples],
+    LL_train,
+    givens={sym_x: sh_x_train[batch_slice]},
+    updates=updates)
+
+test_model = theano.function(
+    [sym_index, sym_batch_size, sym_eq_samples, sym_iw_samples], LL_eval,
+    givens={sym_x: sh_x_test[batch_slice]})
+
+x_sample = (np.random.random((100, latent_size)) > 0.5).astype('float32')
+sample100_model = theano.function([sym_x], x_mu_sample)
+
+
+# Training and Testing functions
+def train_epoch(lr, eq_samples, iw_samples, batch_size):
+    n_train_batches = train_x.shape[0] / batch_size
+    costs = []
+    for i in range(n_train_batches):
+        cost_batch = train_model(i, batch_size, lr, eq_samples, iw_samples)
+        costs += [cost_batch]
+    return np.mean(costs)
+
+
+def test_epoch(eq_samples, iw_samples, batch_size):
+    n_test_batches = test_x.shape[0] / batch_size
+    costs = []
+    output = {'x' :[]}
+    for i in range(n_test_batches):
+        cost_batch = test_model(i, batch_size, eq_samples, iw_samples)
+        costs += [cost_batch]
+    return np.mean(costs), output
+
+print "Training"
+
+# TRAIN LOOP
+# We have made some the code very verbose to make it easier to understand.
+total_time_start = time.time()
+costs_train = []
+costs_eval = []
+LL_test1  = []
+xepochs = []
+for epoch in range(1, 1+num_epochs):
+    start = time.time()
+    np.random.shuffle(train_x)
+    sh_x_train.set_value(preprocesses_dataset(train_x))
+    cost_train_ = train_epoch(lr, eq_samples, iw_samples, batch_size)
+    cost_eval_, output_eval = test_epoch(eq_samples, 1000, batch_size)
+    costs_train += [cost_train_]
+    costs_eval += [cost_eval_]
+    plotsamples(res_out + '/prior.png', sample100_model(
```

Comment:
Yes. I think i removed the sample number because it has to run for several hundred epochs so you'll get a lot of plots :)